### PR TITLE
Better error reporting/catching in agent on nativeRoutingCIDR

### PR DESCRIPTION
**Pull request by jibi**
See individual commits

### CODE REVIEW

**Comment by jibi** [comment]
test-me-please

**Review by qmonnet — APPROVED** [review]
Looks good!

**Comment by jibi** [comment]
test-me-please

**Comment by bmcustodio** [comment]
@jibi could you please rebase now that #16469 has been merged?

**Comment by jibi** [comment]
> @jibi could you please rebase now that #16469 has been merged?

@bmcustodio done :heavy_check_mark: 

I think there are still some issues with the helm documentation as for example the `monitor` key is getting the documentation for `ipv4NativeRoutingCIDR` (since that entire block is commented out)

**Comment by jibi** [comment]
test-me-please

**Comment by bmcustodio** [comment]
> I think there are still some issues with the helm documentation as for example the monitor key is getting the documentation for ipv4NativeRoutingCIDR (since that entire block is commented out)

Thank you :+1: I'm going to check that one out!

File: daemon/cmd/daemon_main.go [hunk 1]
```diff
@@ -624,7 +624,12 @@ func init() {
 	flags.Bool(option.EnableHostFirewall, false, "Enable host network policies (beta when using kube-proxy)")
 	option.BindEnv(option.EnableHostFirewall)
 
-	flags.String(option.IPv4NativeRoutingCIDR, "", "Allows to explicitly specify the CIDR for native routing. This value corresponds to the configured cluster-cidr.")
+	flags.String(option.NativeRoutingCIDR, "",
+		fmt.Sprintf("Allows to explicitly specify the IPv4 CIDR for native routing. This value corresponds to the configured cluster-cidr. Deprecated in favor of --%s", option.IPv4NativeRoutingCIDR))
+	option.BindEnv(option.NativeRoutingCIDR)
+	flags.MarkDeprecated(option.NativeRoutingCIDR, "This option will be removed in v1.12")
```
Comment (borkmann) [inline daemon/cmd/daemon_main.go:631]: nit: should we also add MarkHidden?

**Review by borkmann — APPROVED** [review]
LGTM, thanks!

**Comment by jibi** [comment]
test-me-please

**Comment by jibi** [comment]
retest-runtime

**Comment by jibi** [comment]
retest-4.9

**Review by christarazi — APPROVED** [review]
The 1st and 2nd commit seem to me that they can be squashed. If we ever go back to checking when an option was deprecated, it will be easier if it all happened in one commit.

**Comment by jibi** [comment]
test-me-please

**Review by pchaigno — APPROVED** [review]

**Comment by jibi** [comment]
retest-4.9

**Comment by jibi** [comment]
retest-netnext

**Comment by jibi** [comment]
retest-4.9

**Review by errordeveloper — APPROVED** [review]

**Review by ldelossa — APPROVED** [review]
Looks good to me. 

I question the use of cidr.MustParseCIDR causing a panic while the rest of error conditions in config.go perform a normal logged error and exit the application. But this isn't a change in your PR.

**Comment by jibi** [comment]
> Looks good to me.
> 
> I question the use of cidr.MustParseCIDR causing a panic while the rest of error conditions in config.go perform a normal logged error and exit the application. But this isn't a change in your PR.

Makes sense :+1: perhaps as a follow up we can replace `MustParseCIDR` which something that returns an error rather than panicking (and move the v4 range check inside that function)

**Comment by ldelossa** [comment]
> > Looks good to me.
> > I question the use of cidr.MustParseCIDR causing a panic while the rest of error conditions in config.go perform a normal logged error and exit the application. But this isn't a change in your PR.
> 
> Makes sense  perhaps as a follow up we can replace `MustParseCIDR` which something that returns an error rather than panicking (and move the v4 range check inside that function)

That's my thinkig, but low hanging fruit IMO. Nice to have.

**Comment by jibi** [comment]
retest-4.9

**Comment by borkmann** [comment]
retest-1.21-4.9

**Comment by joestringer** [comment]
@tklauser thanks for keeping an eye out for this, I agree that change should not be made in v1.10 as it can cause upgrade impact or break existing deployments during a point release.

@borkmann @jibi what was the intent here?

I'll drop the backport label until we figure this out.

**Comment by jibi** [comment]
> @tklauser thanks for keeping an eye out for this, I agree that change should not be made in v1.10 as it can cause upgrade impact or break existing deployments during a point release.
> 
> @borkmann @jibi what was the intent here?
> 
> I'll drop the backport label until we figure this out.

@joestringer I marked this for backport as I thought the upgrade impact was minimal (i.e. only a warning that `--native-routing-cidr` is deprecated and that `--ipv4-native-routing-cidr` should be used instead).

What I just realized is that this has a non trivial _downgrade_ impact: if for example a user starts on 1.10.2 following the documentation (using the new `--ipv4-native-routing-cidr`) and then downgrades to 1.10.1, the downgrade will break their setup as the flag does not exist in that release. So I agree with the decision of dropping the label (sorry for the noise)

**Comment by joestringer** [comment]
FWIW taking a look back over, while this could potentially catch some misconfiguration from users in a few circumstances, I don't think this PR really reaches the threshold where we should be backporting it. There is a cost of performing the backport and some degree of risk we take on. In this case I think it's OK for users to pick up these improvements with the next release.

For reference, the backport criteria definition is here: https://docs.cilium.io/en/latest/contributing/release/backports/#backport-criteria-for-current-minor-release.

**Pull request merged**